src/components/AutoScroller/AutoScroller.tsx

```diff
@@ -0,0 +1,140 @@
+import * as React from 'react';
+import noop from 'lodash/noop';
+import { getScrollParent } from './util';
+
+export type Props = Partial<React.Attributes> & {
+    as?: React.ElementType;
+    children?: React.ReactNode;
+    className?: string;
+    enabled?: boolean;
+    onScroll?: (x: number, y: number) => void;
+};
+
+const SCROLL_DELTA = 0.2; // Intensity to ramp up scroll speed as the cursor moves deeper into the scroll gutter
+const SCROLL_GUTTER = 50; // Invisible zone within the component that triggers auto-scroll
+
+export function AutoScroller(props: Props, ref: React.Ref<Element>): JSX.Element {
+    const { as: Element = 'div', children, enabled, onScroll = noop, ...rest } = props;
+
+    // Create refs to store un-rendered data, references, and state
+    const handleRef = React.useRef<number | null>(null);
+    const parentRef = React.useRef<Element | null>(null);
+    const parentRectRef = React.useRef<DOMRect | null>(null);
```

Comment:
Does this need to be a separate ref? Can it be derived from `parentRef`?

Reply:
I just added a comment for this at assignment. It's basically a performance optimization so we don't have to query the DOM constantly for a rect that doesn't change between renders.

Reply:
Makes sense 👍